### PR TITLE
Adjust Jira plugin to return annotation

### LINKED ISSUE #42: Append IssueId to ValidationResponse
### TL;DR

Append IssueId to ValidationResponse

### Detailed design

_No response_

### Alternatives considered

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by qiaoweiz**
- IssueId is returned in the MatchIssue method. 
https://developer.atlassian.com/cloud/jira/platform/rest/v3/api-group-issue-search/#api-rest-api-3-jql-match-post

- fix #42

### CODE REVIEW

File: pkg/plugin/plugin.go [hunk 1]
```diff
@@ -84,10 +85,21 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
...
 
+	// There is only one JQL and one issueKey, so the first match result is
+	// checked directly.
+	if len(result.Matches[0].MatchedIssues) == 1 {
+		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+		return &jvspb.ValidateJustificationResponse{
+			Valid:   true,
+			Warning: result.Matches[0].Errors,
+			Annotation: map[string]string{
+				"jira_issue_id": issueID,
```
Comment (yolocs) [inline pkg/plugin/plugin.go:96]: I thought we are going to add the URL here as well? Or you plan to do that separately

Comment (yolocs) [inline pkg/plugin/plugin.go:96]: I thought we are going to add the URL here as well? Or you plan to do that separately

Comment (qiaoweiz) [reply pkg/plugin/plugin.go]: (same)

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -84,10 +85,24 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
 		}, fmt.Errorf("failed to validate justification: %w", err)
 	}
 
+	// There is only one JQL and one issueKey, so the first match result is
+	// checked directly.
+	if len(result.Matches[0].MatchedIssues) == 1 {
+		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+		baseURL := "https://verily-okta-sandbox.atlassian.net/browse/"
```
Comment (yolocs) [inline pkg/plugin/plugin.go:92]: This should be a config.

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -84,10 +85,24 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
...
+		issueURL := baseURL + req.Justification.Value
+		return &jvspb.ValidateJustificationResponse{
+			Valid:   true,
+			Warning: result.Matches[0].Errors,
+			Annotation: map[string]string{
+				"jira_issue_id":  issueID,
+				"jira_issue_url": issueURL,
```
Comment (yolocs) [inline pkg/plugin/plugin.go:99]: Make the keys consts. Also I wonder if we can save "jira_" since it's already assumed in the context.

**Review by yolocs — CHANGES_REQUESTED** [review]

File: pkg/plugin/plugin.go [hunk 4]
```diff
@@ -84,10 +85,24 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
 		}, fmt.Errorf("failed to validate justification: %w", err)
 	}
 
+	// There is only one JQL and one issueKey, so the first match result is
+	// checked directly.
+	if len(result.Matches[0].MatchedIssues) == 1 {
```
[the remark this answers is not in the document]
Comment (qiaoweiz) [reply pkg/plugin/plugin.go:98]: I think it is part of validator so I adjusted the validator to throw the non-found error.

File: pkg/plugin/plugin.go (hunk 2 above)
Comment (qiaoweiz) [reply pkg/plugin/plugin.go]: Update the config to have a "baseURL"

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (qiaoweiz) [reply pkg/plugin/plugin.go]: I made them consts. As I observed that all the items in the config start with "jira_", I have retained the "jira_" here to maintain consistency.

File: pkg/plugin/config.go [hunk 5]
```diff
@@ -134,5 +141,13 @@ func (cfg *PluginConfig) ToFlags(set *cli.FlagSet) *cli.FlagSet {
 		Usage:   "Hint is for what value to put as the justification.",
 	})
 
+	f.StringVar(&cli.StringVar{
+		Name:    "jira-plugin-base-url",
+		Target:  &cfg.BaseURL,
+		EnvVar:  "JIRA_PLUGIN_BASE_URL",
+		Example: "https://verily-okta-sandbox.atlassian.net/browse/",
```
Comment (sethvargo) [inline pkg/plugin/config.go:148]: Does this actually require the trailing slash?

File: pkg/plugin/plugin.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sethvargo) [reply pkg/plugin/plugin.go:98]: It's usually best to have the defensive checks as close to the code as possible. That way, future refactors don't accidentally break undocumented invariants.

[the remark this answers is not in the document]
Comment (yolocs) [reply pkg/plugin/plugin.go:98]: +1

File: pkg/plugin/plugin.go [hunk 6]
```diff
@@ -84,10 +93,23 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
...
+		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+		issueURL := j.baseURL + req.Justification.Value
```
Comment (yolocs) [inline pkg/plugin/plugin.go:100]: We would need to add the slash here: `j.baseURL + "/" + req.Justification.Value`

**Review by yolocs — CHANGES_REQUESTED** [review]

File: pkg/plugin/config.go [hunk 7]
```diff
@@ -134,5 +141,13 @@ func (cfg *PluginConfig) ToFlags(set *cli.FlagSet) *cli.FlagSet {
 		Usage:   "Hint is for what value to put as the justification.",
 	})
 
+	f.StringVar(&cli.StringVar{
+		Name:    "jira-plugin-base-url",
+		Target:  &cfg.BaseURL,
+		EnvVar:  "JIRA_PLUGIN_BASE_URL",
```
[the remark this answers is not in the document]
Comment (qiaoweiz) [reply pkg/plugin/config.go]: Updated in latest version.

File: pkg/plugin/plugin.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (qiaoweiz) [reply pkg/plugin/plugin.go:103]: Added a empty check.

File: pkg/plugin/config.go (hunk 5 above)
Comment (qiaoweiz) [reply pkg/plugin/config.go]: Updated the the example URL

[the remark this answers is not in the document]
Comment (qiaoweiz) [reply pkg/plugin/config.go]: Removed "verily"

File: pkg/plugin/plugin.go (hunk 6 above)
Comment (qiaoweiz) [reply pkg/plugin/plugin.go]: Update the issueURL as url.JoinPath(j.issueBaseURL, "/browse/", req.Justification.Value)

**Review by yolocs — CHANGES_REQUESTED** [review]
nits

File: pkg/plugin/plugin.go [hunk 8]
```diff
@@ -84,10 +94,32 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
...
+	if len(result.Matches) == 0 || len(result.Matches[0].MatchedIssues) == 0 {
+		return nil, fmt.Errorf("failed to get the matched jira issue %q", req.Justification.Value)
+	}
+
+	// There is only one JQL and one issueKey, so the first match result is
+	// checked directly.
+	if len(result.Matches[0].MatchedIssues) == 1 {
+		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+
+		issueURL, err := url.JoinPath(j.issueBaseURL, "/browse/", req.Justification.Value)
```
[the remark this answers is not in the document]
Comment (yolocs) [reply pkg/plugin/plugin.go:107]:
```suggestion
		issueURL, err := url.JoinPath(j.issueBaseURL, "browse", req.Justification.Value)
```
We don't need slashes if we already use JoinPath?

File: pkg/plugin/plugin.go [hunk 9]
```diff
@@ -84,10 +94,33 @@ func (j *JiraPlugin) Validate(ctx context.Context, req *jvspb.ValidateJustificat
...
+	if len(result.Matches) == 0 || len(result.Matches[0].MatchedIssues) == 0 {
+		return nil, fmt.Errorf("failed to get the matched jira issue %q", req.Justification.Value)
+	}
+
+	// There is only one JQL and one issueKey, so the first match result is
+	// checked directly.
+	if len(result.Matches[0].MatchedIssues) == 1 {
+		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+
+		// The format for the Jira issue URL follows the pattern "thttps://<user’s subdomain>.atlassian.net/browse/<issueKey>".
```
Comment (yolocs) [inline pkg/plugin/plugin.go:106]:
```suggestion
		// The format for the Jira issue URL follows the pattern "https://your-domain.atlassian.net/browse/<issueKey>".
```
Do you have any official jira reference doc?

**Review by yolocs — APPROVED** [review]

**Pull request merged**